hyperhtml.js

```diff
@@ -193,321 +156,692 @@ var hyperHTML = (function () {'use strict';
+            node.parentNode.insertBefore(value, node);
           }
           break;
       }
     };
   }
 
-  // basic closure to update nodes textContent
-  //
-  // render`
-  //  <p>
-  //    ${'spaces around means textContent'}
-  //  </p>`;
-  function setTextContent(node) {
-    var oldValue;
-    return function text(newValue) {
-      if (oldValue !== newValue) {
-        node.textContent = (oldValue = newValue);
+  // ---------------------------------------------
+  // DOM Traversing
+  // ---------------------------------------------
+
+  // look for attributes that contains the comment text
+  function attributesSeeker(node, paths) {
+    for (var
+      attribute,
+      value = IE ? UID : UIDC,
+      attributes = node.attributes,
+      i = 0, length = attributes.length;
+      i < length; i++
+    ) {
+      attribute = attributes[i];
+      if (attribute.value === value) {
+        paths.push(
+          Path(
+            'attr',
+            // with IE the order doesn't really matter
+            // as long as the right attribute is addressed
+            IE ?
+              node[GET_ATTRIBUTE_NODE](IEAttributes.shift()) :
+              attribute
+          )
+        );
       }
-    };
+    }
   }
 
+  // walk the fragment tree in search of comments
+  function commentsSeeker(node, paths) {
+    for (var
+      child, text,
+      childNodes = node.childNodes,
+      length = childNodes.length,
+      i = 0; i < length; i++
+    ) {
+      child = childNodes[i];
+      switch (child.nodeType) {
+        case ELEMENT_NODE:
+          attributesSeeker(child, paths);
+          commentsSeeker(child, paths);
+          break;
+        case COMMENT_NODE:
+          if (child.textContent === UID) {
+            if (length === 1) {
+              paths.push(Path('any', node));
+            } else if (
+              (i < 1 || childNodes[i - 1].nodeType === ELEMENT_NODE) &&
+              (i + 1 === length || childNodes[i + 1].nodeType === ELEMENT_NODE)
+            ) {
+              paths.push(Path('virtual', child));
+            } else {
+              text = createText(child, '');
+              child.parentNode.replaceChild(text, child);
+              paths.push(Path('text', text));
+            }
+          }
+          break;
+        case TEXT_NODE:
+          if (SHOULD_USE_ATTRIBUTE.test(node.nodeName) && child.textContent === UIDC) {
+            paths.push(Path('text', node));
+          }
+          break;
+      }
+    }
+  }
 
-  // -------------------------
+  // ---------------------------------------------
+  // Constants
+  // ---------------------------------------------
+
+  // Node.CONSTANTS
+  // without assuming Node is globally available
+  // since this project is used on the backend too
+  var ELEMENT_NODE = 1;
+  var ATTRIBUTE_NODE = 2;
+  var TEXT_NODE = 3;
+  var COMMENT_NODE = 8;
+  var DOCUMENT_FRAGMENT_NODE = 11;
+
+  var GET_ATTRIBUTE_NODE = 'getAttributeNode';
```

Comment:
plays well with minifiers and it's used as constant in switch cases too. Now locally I also have `OWN_SVG_OWNER` with similar role.